### PR TITLE
feat(addMetadata): added creationdate,lastModifiedDate,usernameCreate…

### LINKED ISSUE #1477: Add more metadata to candidate licenses
### Description

Add metadata to candidate licenses
* date of creation
* user name created
* user name last modified
* date of last modified

#### Steps followed and expected result

Add this as metadata to get more understand who created these records.

#### Additonal 1

Please rename all UI occurrences to "Candidate Licenses". Sometime it is called "Advice Licenses" which is confusing.

#### Additonal 2

Existing licenses should be inited with fossy (general admin account)

### PR DESCRIPTION
**Pull request by shivamgoyal7**
## Description

Added metadata in Candidatelicense page
1. Creation date
2. Modified date
3. username created
4. username modified

### Changes

added metadata on the candidate license page
Change page name of advice license to the candidate license
Added default username created is fossy

## How to test
start PostgreSQL and apache 2
build and install src
create new license 

Closes  #1477

### CODE REVIEW

**Comment by GMishx** [comment]
Hello @shivamgoyal7 , Thank you for your contribution. The tests are failing for your branch.
Maybe adding `createInheritedTables(['license_candidate'])` before `alterTables()` in `src/testing/db/createTestEnvironment.php` could help.

image

File: src/lib/php/Test/TestPgDb.php [hunk 1]
```diff
@@ -178,7 +178,7 @@ public function createInheritedTables($tableList=array())
   {
     $table = 'license_candidate';
     if ((empty($tableList) || in_array($table, $tableList)) && !$this->dbManager->existsTable($table)) {
-      $this->dbManager->queryOnce("CREATE TABLE $table (group_fk integer) INHERITS (license_ref)");
+      $this->dbManager->queryOnce("CREATE TABLE $table (group_fk integer,rf_creationdate timestamptz,rf_lastmodified timestamptz,rf_user_fk_created text,rf_user_fk_modified text) INHERITS (license_ref)");
```
Comment (GMishx) [inline src/lib/php/Test/TestPgDb.php:181]: Shouldn't `rf_user_fk_created` and `rf_user_fk_modified` be integers and not text to match the `core-schema.dat`?

File: src/www/ui/core-schema.dat [hunk 2]
```diff
@@ -1006,6 +1006,22 @@
...
+  $Schema["TABLE"]["license_candidate"]["rf_lastmodified"]["ADD"] = "ALTER TABLE \"license_candidate\" ADD COLUMN \"rf_lastmodified\" timestamptz DEFAULT now()";
+  $Schema["TABLE"]["license_candidate"]["rf_lastmodified"]["ALTER"] = "ALTER TABLE \"license_candidate\" ALTER COLUMN \"rf_lastmodified\" DROP NOT NULL, ALTER COLUMN \"rf_lastmodified\" SET DEFAULT now()";
+
+  $Schema["TABLE"]["license_candidate"]["rf_user_fk_created"]["DESC"] = "COMMENT ON COLUMN \"license_candidate\".\"rf_user_fk_created\" IS 'group seeing this candidate'";
+  $Schema["TABLE"]["license_candidate"]["rf_user_fk_created"]["ADD"] = "ALTER TABLE \"license_candidate\" ADD COLUMN \"rf_user_fk_created\" int4 DEFAULT 3";
+  $Schema["TABLE"]["license_candidate"]["rf_user_fk_created"]["ALTER"] = "ALTER TABLE \"license_candidate\" ALTER COLUMN \"rf_user_fk_created\" DROP NOT NULL";
+
+  $Schema["TABLE"]["license_candidate"]["rf_user_fk_modified"]["DESC"] = "COMMENT ON COLUMN \"license_candidate\".\"rf_user_fk_modified\" IS 'group seeing this candidate'";
```
Comment (GMishx) [inline src/www/ui/core-schema.dat:1022]: Please update the `DESC` (comment) for the added columns.

File: src/www/ui/page/AdviceLicense.php [hunk 3]
```diff
@@ -163,19 +174,12 @@ private function saveInput(Request $request, $oldRow)
...
-    return array('rf_pk' => $oldRow['rf_pk'],
-        'rf_shortname' => $shortname,
-        'rf_fullname' => $fullname,
-        'rf_text' => $rfText,
-        'rf_url' => $url,
-        'rf_notes' => $note,
-        'rf_risk' => $riskLvl,
-        'marydone' => $marydone);
+      $rfText, $url, $note, $lastmodified, $userIdmodified, !empty($marydone), $riskLvl);
+      return $this->getDataRow(Auth::getGroupId(), $oldRow['rf_pk']);
```
Comment (GMishx) [inline src/www/ui/page/AdviceLicense.php:182]: Please fix the indentation.

```suggestion
    return $this->getDataRow(Auth::getGroupId(), $oldRow['rf_pk']);
```

File: src/www/ui/template/advice_license-edit.html.twig [hunk 4]
```diff
@@ -12,6 +12,24 @@
   <input type="hidden" name="mod" value="advice_license"/>
   <input type="hidden" name="rf" value="{{ rf_pk }}"/>
   <table>
+  {% if rf_lastmodified is not empty %}
+    <tr>
+      <td>{{ 'Username Created'|trans }}:</td>
```
Comment (GMishx) [inline src/www/ui/template/advice_license-edit.html.twig:17]: Just thinking, "User Created" is better than "Username Created" while displaying in UI. What do you think?

File: src/www/ui/template/advice_license-edit.html.twig [hunk 5]
```diff
@@ -12,6 +12,24 @@
...
+      <td><span type="input" name="useridcreated"> {{ rf_user_fk_created|e }}</span></td>
+    </tr>    
+    <tr>
+      <td>{{ 'Date of Creation'|trans }}:</td>
+      <td><span type="input" name="creationdate"> {{ rf_creationdate|e }}</span></td>
+    </tr>
+    <tr>
+      <td>{{ 'Username Modified'|trans }}:</td>
```
Comment (GMishx) [inline src/www/ui/template/advice_license-edit.html.twig:25]: If you are changing above, please consider "User Modified" here as well.

File: src/www/ui/template/advice_license-edit.html.twig [hunk 6]
```diff
@@ -12,6 +12,24 @@
   <input type="hidden" name="mod" value="advice_license"/>
   <input type="hidden" name="rf" value="{{ rf_pk }}"/>
   <table>
+  {% if rf_lastmodified is not empty %}
+    <tr>
+      <td>{{ 'Username Created'|trans }}:</td>
+      <td><span type="input" name="useridcreated"> {{ rf_user_fk_created|e }}</span></td>
```
Comment (GMishx) [inline src/www/ui/template/advice_license-edit.html.twig:18]: As per standard HTML, `<span>` tag does not have `type` and `name` attributes. Can you please either remove them or make it an `<input type="text">` element with `disabled` attribute?

**Review by GMishx — CHANGES_REQUESTED** [review]
Please consider following suggestions.

File: src/lib/php/Test/TestPgDb.php (hunk 1 above)
Comment (shivamgoyal7) [reply src/lib/php/Test/TestPgDb.php:181]: okay

File: src/www/ui/core-schema.dat (hunk 2 above)
Comment (shivamgoyal7) [reply src/www/ui/core-schema.dat:1022]: yeah okay

File: src/www/ui/page/AdviceLicense.php (hunk 3 above)
Comment (shivamgoyal7) [reply src/www/ui/page/AdviceLicense.php:182]: done

File: src/www/ui/template/advice_license-edit.html.twig (hunk 4 above)
Comment (shivamgoyal7) [reply src/www/ui/template/advice_license-edit.html.twig:17]: yeah

File: src/www/ui/template/advice_license-edit.html.twig (hunk 5 above)
Comment (shivamgoyal7) [reply src/www/ui/template/advice_license-edit.html.twig:25]: yeah sure

File: src/www/ui/template/advice_license-edit.html.twig (hunk 6 above)
Comment (shivamgoyal7) [reply src/www/ui/template/advice_license-edit.html.twig:18]: yeah okay

File: src/lib/php/Test/TestPgDb.php (hunk 1 above)
Comment (shivamgoyal7) [reply src/lib/php/Test/TestPgDb.php:181]: so I have to change its data type?

Comment (GMishx) [reply src/lib/php/Test/TestPgDb.php:181]: Yes, the test data should match as closely as possible to production in order to provide optimum test coverage.

Comment (shivamgoyal7) [reply src/lib/php/Test/TestPgDb.php]: i have changed it to varchar

Comment (GMishx) [reply src/lib/php/Test/TestPgDb.php]: Okay but the core-schema says it should be int4????

**Review by GMishx — APPROVED** [review]
Changes looks good. Need testing

**Comment by GMishx** [comment]
image

I guess that's not expected....

**Comment by shivamgoyal7** [comment]
I have done the required changes.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**